test/test_torch.py

```diff
@@ -4891,6 +4891,136 @@ def test_diff(self, device, dtype):
             scalar = torch.tensor(2, device=device, dtype=dtype)
             torch.diff(scalar)
 
+    def _test_histogram_numpy(self, t, bins, bin_range, weights, density):
+        def to_np(t):
+            if not torch.is_tensor(t):
+                return t
+            if t.dtype == torch.bfloat16:
+                return t.to(dtype=torch.float, device="cpu").numpy()
+            else:
+                return t.cpu().numpy()
+
+        (minv, maxv) = bin_range if bin_range else (None, None)
+        (actual_hist, actual_bin_edges) = torch.histogram(t, bins, min=minv, max=maxv, weight=weights, density=density)
+
+        actual_dtype = actual_hist.dtype
+
+        # Helper for test_histogram to compare with NumPy reference implementation
+        def reference_histogram(self, t, bins, bin_range, weights, density):
+            (np_t, np_bins, np_weights) = map(to_np, [t, bins, weights])
+            (np_hist, np_bin_edges) = np.histogram(np_t, np_bins, range=bin_range, weights=np_weights, density=density)
+            return (torch.from_numpy(np_hist).to(actual_dtype), torch.from_numpy(np_bin_edges).to(actual_dtype))
+
+        (expected_hist, expected_bin_edges) = reference_histogram(self, t, bins, bin_range, weights, density)
+
+        """
+        Works around linspace discrepancies by passing torch's constructed bin_edges to numpy.
+
+        When bin edges are not explicitly defined, histogram uses the linspace operator internally
+        to construct the sequence of bin edges. In some cases, torch.linspace output differs slightly
+        from numpy.linspace output.
+        Issue: https://github.com/pytorch/pytorch/issues/58758
+        """
+        if not torch.is_tensor(bins):
+            self.assertEqual(actual_bin_edges, expected_bin_edges, atol=1e-5, rtol=1e-5)
+            # Calls numpy.histogram again, passing torch's actual_bin_edges as the bins argument
+            (expected_hist, expected_bin_edges) = reference_histogram(self, t, actual_bin_edges, bin_range, weights, density)
+
+        self.assertEqual(actual_hist, expected_hist)
+        self.assertEqual(actual_bin_edges, expected_bin_edges)
+
+    @onlyCPU
+    @dtypes(torch.float32, torch.float64)
+    def test_histogram(self, device, dtype):
```

Comment:
Is it also interesting to test handwritten edge cases like a tensor where all the values are equal? Or where min=max? Or where there are values exactly equal to all the edges?